File: packages/sdkwork-utils-go/crypto.go
package utils

import (
	"crypto/hmac"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
)

func Sha256Hash(value []byte) string {
	digest := sha256.Sum256(value)
	return hex.EncodeToString(digest[:])
}

func HmacSha256(value []byte, secret []byte) string {
	mac := hmac.New(sha256.New, secret)
	_, _ = mac.Write(value)
	return hex.EncodeToString(mac.Sum(nil))
}

func SecureCompare(left, right string) bool {
	if len(left) != len(right) {
		return false
	}
	return subtle.ConstantTimeCompare([]byte(left), []byte(right)) == 1
}

File: packages/sdkwork-utils-typescript/src/crypto.ts
import { hmacSha256Hex, sha256Hex } from "./runtime/sha256.js";

export function sha256Hash(value: string | Uint8Array): string {
  return sha256Hex(value);
}

export function hmacSha256(value: string | Uint8Array, secret: string | Uint8Array): string {
  return hmacSha256Hex(value, secret);
}

export function secureCompare(left: string, right: string): boolean {
  if (left.length !== right.length) {
    return false;
  }
  let result = 0;
  for (let index = 0; index < left.length; index += 1) {
    result |= left.charCodeAt(index) ^ right.charCodeAt(index);
  }
  return result === 0;
}
